room/runtime/src/main/java/androidx/room/RoomDatabase.java

```diff
@@ -1012,6 +1058,21 @@ private static boolean isLowRamDevice(@NonNull ActivityManager activityManager)
             return this;
         }
 
+        /**
+         * Adds a {@link TypeConverterFactory} to this database.
+         *
+         * @param factory The factory.
+         * @return This {@link Builder} instance.
+         */
+        @NonNull
+        public Builder<T> addTypeConverterFactory(@NonNull TypeConverterFactory factory) {
+            if(mTypeConverterFactories == null) {
+                mTypeConverterFactories = new HashMap<>();
+            }
+            mTypeConverterFactories.put(factory.getClass().getCanonicalName(), factory);
```

Comment:
One interesting issue here is that we build the map with the concrete factory class, so if a user provides a subclass of an expected factory Room won't be able to correctly use it / accept it. One solution to this is processing the factory type and making sure its `final`, i.e. making it a compiler error if its not. This is a bit aggressive IMHO, but would avoid any confusion since we can't quite support it.

Reply:
Good point, I believe Yigit mentioned this as well. I can add a `final` check but I'm trying to understand why can't we support this. Wouldn't it be possible to do something like this:
1. Try to find the converter in the map and use it if it's there
2. Otherwise try to go over supertypes of a given factory and for every one of them check if it's in the map.
Not sure if it's worth it though.

Reply:
Right, we can't do this without going through superclasses at runtime via reflection. My take here is, it's not initially worth it, I would prefer if users compose their factories (and delegate) not via class hierarchy. We can always add support for this later (and remove the `final` restriction) if we get users with compelling reasons, but if we don't make it strict now and decided not to support this case in the long run we can't later add the restriction since it would cause ABI incompatibilities.

Reply:
Makes sense, I'll add `final` check

Reply:
Do you think we should add an info about `final` requirement in the javadoc for `TypeConverterFactory`  interface?

Reply:
Yes, definitely